specs/Popup-AdditionalLayoutProperties-Spec.md

```diff
@@ -0,0 +1,218 @@
+
+Secondary commands flyout closed:
+
+![Shows the CommandBarFlyout not yet open with insufficient monitor space below it](images/CommandBarFlyout-InsufficientSpace.png)
+
+Secondary commands flyout open:
+
+![Shows the CommandBarFlyout opened up](images/CommandBarFlyout-InsufficientSpace-Open.png)
+
+# API Pages
+
+## Popup class
+
+```csharp
+class Popup
+{
+    // Existing APIs
+    // ...
+
+    // New APIs
+    UIElement PlacementTarget { get; set; }
+    PopupPlacementMode DesiredPlacement { get; set; }
+    PopupPlacementMode ActualPlacement { get; }
+    
+    public event System.EventHandler<object> PlacementChanged;
+}
+```
+
+The example below shows how the `Placement` APIs are used by the 
+[CommandBarFlyout](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.CommandBarFlyout)
+to position its
+[CommandBarFlyoutCommandBar](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.Primitives.CommandBarFlyoutCommandBar)'s
+secondary commands Popup, and how to respond to the event raised when XAML places the Popup.
+
+```xml
+<!-- Part of the CommandBarFlyoutCommandBar's default template -->
+<Popup
+    x:Name="OverflowPopup"
+    PlacementTarget="{Binding ElementName=PrimaryItemsRoot}"
+    DesiredPlacement="Bottom">
+</Popup>
+```
+
+```csharp
+void OnApplyTemplate()
+{
+    m_overflowPopup = GetTemplateChild("OverflowPopup");
+    m_overflowPopup.PlacementChanged += OnOverflowPopupPlacementChanged;
+}
+
+void OnOverflowPopupPlacementChanged(object sender, object args)
+{
+    UpdateVisualState(useTransitions: false);
+}
+
+void UpdateVisualState(bool useTransitions)
+{
+    if (m_overflowPopup.ActualPlacement == PopupPlacementMode.Top)
+    {
+        VisualStateManager.GoToState(this, "ExpandedUp", useTransitions);
+    }
+    else
+    {
+        VisualStateManager.GoToState(this, "ExpandedDown", useTransitions);
+    }
+}
+```
+
+## Popup.PlacementTarget property
+
+Use this property to describe which element the `Popup` should be positioned relative to.
+Defaults to `null`.
+
+If this is `null`, then `DesiredPlacement` is ignored, `ActualPlacement` is always `None`, and
+`PlacementChanged` is never raised.  If the `Popup` is in the visual tree, `PlacementTarget` will override what its
+position would otherwise be set to by layout.  Setting `PlacementTarget` to an element under a different XAML root than
+`Popup.XamlRoot` is invalid and will throw an `ArgumentException`.
+
+_Spec note: this property is analogous to the
+[FlyoutBase.Target](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.Primitives.FlyoutBase.Target)
+property, but `Popup.Target` looked confusing, so we added the `Placement` prefix._
+
+## Popup.DesiredPlacement property
+
+Use this property to describe how the you would ideally like the `Popup`
+positioned relative to `PlacementTarget`.  Defaults to `None`.
+
+If this is `None`, then `PlacementTarget` is ignored,
+`ActualPlacement` is always `None` and `PlacementChanged` is never raised. 
```

Comment:
This seems wrong. If `PlacementTarget` is non-null, then we should still get an `ActualPlacement` to tell us where it got placed relative to the target, even if I expressed no preference. For example, suppose CommandBar didn't really care which way the popup went, but it did want to know where it ended up, so it could draw its collapse button arrow in the correct direction.

Reply:
If PreferredPlacement has its default value of None, then the popup isn't placed relative to PlacementTarget - it has its default behavior of being placed at an absolute position if it's not in the visual tree, or being placed by layout if it is.  The idea is that the two properties only take effect if both have been set.

Reply:
It's confusing that when Popup doesn't have a PlacementTarget, it still sort of of does if it's in the tree (it's a parented popup). Then it gets placed relative to the parent (aligned to the top/left corner of the parent?)

The DesiredPlacement only has affect if PlacementTarget is set, somehow we need to align the properties, make it evident that they go together. Like `DesiredPlacementRelativeToTarget`.

Reply:
It has a *placement* if it's in the tree, but it doesn't really have a target that it's placed *against*.  It just goes wherever an element would go according to layout.  So, for example, if you had this:

```xml
<StackPanel>
  <CheckBox />
  <Popup />
  <CheckBox />
</StackPanel>
```

...then the Popup would be positioned between the two CheckBoxes, same as if it were another CheckBox.  There isn't really anything you could call a singular placement target in this case.

This is one of the reasons I kind of like the combination of `Placement`/`PlacementTarget`; that says to me "this is the placement we want, and this is the target we want for that placement".  Though the point is still entirely valid regarding the `Desired` prefix that we might not place it at the placement specified.

I can't think of a non-wordy way to specify that the two properties go together while also keeping the "try"-esque semantics behind `DesiredPlacement`.